docs/proposals/12-oidc-webhook-authenticator.md

```diff
@@ -0,0 +1,305 @@
+
+    ```json
+    {
+      "apiVersion": "authorization.k8s.io/v1",
+      "kind": "SubjectAccessReview",
+      "spec": {
+        "groups": [
+          "system:serviceaccounts",
+          "system:serviceaccounts:shoot--abcd",
+          "system:authenticated"
+        ],
+        "nonResourceAttributes": {
+          "path": "/validate-token",
+          "verb": "post"
+        },
+        "user": "system:serviceaccount:shoot--abcd:kube-apiserver"
+      },
+      "status": {
+        "allowed": true,
+        "reason": "RBAC: allowed by RoleBinding \"kube-apiserver\" of ClusterRole \"kube-apiserver\" to ServiceAccount \"system:serviceaccount:shoot--abcd:kube-apiserver\""
+      }
+    }
+    ```
+
+1. OWA then iterates over all registered `OpenIDConnect` and tries to validate the token.
+1. Upon a successful validation it returns the `TokeReview` with user, groups and extra parameters:
+
+    ```json
+    {
+      "TokenReview": {
+        "kind": "TokenReview",
+        "apiVersion": "authentication.k8s.io/v1beta1",
+        "spec": {
+          "token": "ddeewfwef..."
+        },
+        "status": {
+          "authenticated": true,
+          "user": {
+            "username": "test-foo@bar.com",
+            "groups": [
+              "baz-employee"
+            ],
+            "extra": {
+              "gardener.cloud/apiserver/groups": [
+                "system:serviceaccounts",
+                "system:serviceaccounts:shoot--abcd",
+                "system:authenticated"
+              ],
+              "gardener.cloud/apiserver/uid": [
+                "system:serviceaccount:shoot--abcd:kube-apiserver"
+              ],
+              "gardener.cloud/apiserver/username": [
+                "system:serviceaccount:shoot--abcd:kube-apiserver"
+              ],
+              "gardener.cloud/oidc/name": [
+                "foo"
+              ],
+              "gardener.cloud/oidc/uid": [
+                "e5062528-e5a4-4b97-ad83-614d015b0979"
+              ]
+            }
+          }
+        }
+      }
+    }
+    ```
+
+    It also adds some extra information which can be used by custom authorizers later on:
+
+    1. `gardener.cloud/apiserver/groups` contains all the groups of the API server which is making the `TokenReview` request (it's the ServiceAccount of the API Server Pod in this case)
+    1. `gardener.cloud/apiserver/uid` contains the UID of the API server which is making the `TokenReview` request (it's the ServiceAccount of the API Server Pod in this case)
+    1. `gardener.cloud/apiserver/username` contains the username of the API server which is making the `TokenReview` request (it's the ServiceAccount of the API Server Pod in this case)
+    1. `gardener.cloud/oidc/name` contains the name of the `OpenIDConnect` authenticator which was used.
+    1. `gardener.cloud/oidc/uid` contains the UID of the `OpenIDConnect` authenticator which was used.
```

Comment:
How should this UID be generated?  Possibly a hash over the `OpenIDConnect` CRD plus project name?

Reply:
`metadata.uid` of the `OpenIDConnect`

Reply:
I'll add `resourceVersion` as well.